### PR TITLE
Themes: Translate custom templates in theme.json

### PR DESCRIPTION
**Pull request by gziolo**
## Description

An alternative to #28783. The difference is that rather than changing the way translatable fields are marked in `experimental-18n-theme.json`, the structure of `experimental-theme.json` for `customTemplates` and `templateParts` gets updated to fit the i18n extracting logic. It also aligns with how settings are declared like color palettes for font sizes.

**Before**

```json
{
    "templateParts": {
        "header": { "area": "header" },
        "footer": { "area": "footer" }
    }
}
```

**After**

```json
{
    "templateParts": [
        {
           "name":  "header",
            "area": "header"
        },
       {
           "name":  "footer",
           "area": "footer"
        }
    ]
}
```

https://github.com/WordPress/gutenberg/pull/27778 introduced a new field in the `theme.json` file, the `customTemplates` to declare custom page templates to be shown to the user in places like:

Captura de ecrã de 2021-02-04 14-21-55

https://github.com/WordPress/gutenberg/pull/28700 integrated the processing of that field into the `WP_Theme_JSON` / `WP_Theme_JSON_Resolver` classes. This is the next step of that integration: make it translatable.

Changes:

- [x] Added a new in `experimental-i18n-theme.json`, so wp-cli can understand what needs to be translated.
- [x] Update `WP_Theme_JSON_Resolver` to retrieve the translated version of this field.

## How has this been tested?

1. I activated `tt1-blocks` theme.
2. Updated `experimental-theme.json` file to contain:
  ```json
	"customTemplates": [
		{
			"name": "page-home",
			"title": "Page without title"
		}
	]
  ```
3. Added the following hook to test if translation gets applied:
  ```php
add_filter( 'gettext_with_context', function ( $translation, $text, $context, $domain ) {
	if ( 'tt1-blocks' === $domain && 'Page without title' === $text ) {
		return 'Strona bez tytułu';
	}
	return $translation;
}, 10, 4 );
  ```
4. Opened a page in the block editor and checked if the custom template is translated in the combobox in the Page Attributes section.

In addition to that, a new unit test was added to cover translation logic applied to `theme.json`.

## Screenshots

Screen Shot 2021-03-15 at 15 32 59
Screen Shot 2021-03-15 at 15 33 19

## Types of changes
New feature (non-breaking change which adds functionality).

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: phpunit/data/themedir1/fse/experimental-theme.json [hunk 1]
```diff
@@ -0,0 +1,33 @@
+{
+	"settings": {
+		"root": {
```
Comment (gziolo) [inline phpunit/data/themedir1/fse/experimental-theme.json:3]: @nosolosw – is it `defaults` or `root` now? According to [docs](https://github.com/WordPress/gutenberg/blob/trunk/docs/how-to-guides/themes/theme-json.md), it's `defaults` but it didn't work in my testing.

**Comment by oandregal** [comment]
According to https://github.com/WordPress/gutenberg/pull/28410#issuecomment-797677594 it looks like `templateAreas` are not translatable. I see we have some references to this: issue description, test fixture (experimental-theme.json) and changes to WP_Theme_JSON (plus the corresponding test). I presume we have to revert those changes before merging this?

**Comment by gziolo** [comment]
> According to [#28410 (comment)](https://github.com/WordPress/gutenberg/pull/28410#issuecomment-797677594) it looks like `templateAreas` are not translatable. I see we have some references to this: issue description, test fixture (experimental-theme.json) and changes to WP_Theme_JSON (plus the corresponding test). I presume we have to revert those changes before merging this?

They aren't translatable today. We can't tell what the future brings but it's better to keep them structured the same way as `customTemplates` or anything that stores lists: font sizes, color palettes, etc.

File: phpunit/data/themedir1/fse/experimental-theme.json (hunk 1 above)
Comment (oandregal) [reply phpunit/data/themedir1/fse/experimental-theme.json:3]: It's fine as it is, I see [the test](https://github.com/WordPress/gutenberg/pull/29828/files#diff-9832681f57d8164557e439d992dca30683e8b387bbdf72817b0e283ac8214de2R110) also uses `root`. I changed both to `defaults` and also works.

**Review by oandregal — APPROVED** [review]
I see there's some conflicts with the theme-json docs. Other than that this is ready.

Special thanks for taking the time to add a new test for this! This should increase our confidence levels by an order of magnitude.

**Comment by oandregal** [comment]
Oh, just remembered one thing: when there're changes to the theme.json format I usually create the corresponding PR in theme-experiments (or let the themers know what changes are upcoming).

**Comment by gziolo** [comment]
> Oh, just remembered one thing: when there're changes to the theme.json format I usually create the corresponding PR in theme-experiments (or let the themers know what changes are upcoming).

Thanks for the reminder. I thought about that when I was testing changes with `tt1-blocks` 😄 I also checked `Q` that uses the old format.

**Comment by gziolo** [comment]
I opened PR for TT1-blocks theme: https://github.com/WordPress/theme-experiments/pull/232
@aristath already update the Q theme.
I also asked on WordPress Slack @fklein-lu and @carolinan to check their FSE themes.

**Review by carolinan — COMMENTED** [review]
With this PR I can't get the `"postTypes": [ "page", "post" ]` setting to work.  The template is only available for pages.

**Comment by gziolo** [comment]
> With this PR I can't get the `"postTypes": [ "page", "post" ]` setting to work. The template is only available for pages.

It's my last refactoring where I made it explicit what is sourced from the config file and I forgot to list `postTypes` 🙈 I will fix it, thank you for testing 👍🏻

**Comment by gziolo** [comment]
> With this PR I can't get the `"postTypes": [ "page", "post" ]` setting to work.  The template is only available for pages.

@carolinan, it should be fixed now.

File: lib/class-wp-theme-json-resolver.php [hunk 2]
```diff
@@ -151,6 +151,34 @@ public static function get_presets_to_translate() {
...
+		}
+
+		foreach ( $array_to_translate as $item_key => $item_to_translate ) {
+			if ( empty( $item_to_translate[ $key ] ) ) {
+				continue;
+			}
+
+			// phpcs:ignore WordPress.WP.I18n.LowLevelTranslationFunction,WordPress.WP.I18n.NonSingularStringLiteralText,WordPress.WP.I18n.NonSingularStringLiteralContext,WordPress.WP.I18n.NonSingularStringLiteralDomain
+			$array_to_translate[ $item_key ][ $key ] = translate_with_gettext_context( $array_to_translate[ $item_key ][ $key ], $context, $domain );
+			// phpcs:enable
```
Comment (swissspidy) [inline lib/class-wp-theme-json-resolver.php]: You don't need `phpcs:enable` here. You would only need it to re-enable after a `phpcs:disable`. But using that combo is rarely needed.

So only the `phpcs:ignore` above should be enough

Comment (gziolo) [reply lib/class-wp-theme-json-resolver.php]: I copied it from another place, so I will update all instances. Thank you for raising it 👍🏻

**Review by aristath — APPROVED** [review]

**Comment by aristath** [comment]
This PR somehow caused an issue with missing items in theme.json files 🤔 
See https://github.com/WordPress/gutenberg/pull/30127 & https://github.com/WordPress/gutenberg/pull/30144
Reverting this fixes the other issues, and the site-editor loads. Commenting-out line https://github.com/WordPress/gutenberg/pull/29828/files#diff-41112e4d1e1b9d7fe224c7c598fb56d7ed165d57335008129e1c4658252fb037R206 also fixes the issue, but I thought @gziolo it might be best if you look at this since chances are you understand it better than I 👍

**Comment by gziolo** [comment]
I think I might introduce a regression there around presets. I will improve our test coverage first and see how to fix it.

**Pull request merged**